src/structures/gdt.rs

```diff
@@ -105,6 +105,25 @@ impl GlobalDescriptorTable {
         }
     }
 
+    /// Forms a GDT from a raw table and a next_free counter.
+    ///
+    /// # Safety
+    ///
+    /// The user must ensure that the provided table contains well formed entries
+    /// and that the value of `next_free` does not exceed the maximum size of the table.
+    #[inline]
+    pub const unsafe fn from_raw_parts(table: [u64; 8], next_free: usize) -> GlobalDescriptorTable {
+        GlobalDescriptorTable { table, next_free }
+    }
+
+    /// Breaks a GDT into its raw parts (table and a next_free counter.)
+    #[inline]
+    pub const fn into_raw_parts(&self) -> ([u64; 8], usize) {
+        let Self { table, next_free } = *self;
+
+        (table, next_free)
+    }
```

Comment:
As above, I would prefer a slice return type here instead of exposing `next_free`.

Reply:
Just to clarify, are you suggesting we return a reference to the inner table coerced to a slice directly?

Reply:
Sorry, let me give you a few more details. The `GlobalDescriptorTable` type is basically a fixed-capacity `Vec`. The `next_free` field is similar to the `len` property of a `Vec`. Returning the unused entries of the GDT doesn't make much sense in my opinion, so I would prefer to return a `&[u64]` slice constructed from `&self.table[..self.next_free]` instead.

Reply:
I was worried that you were thinking of encoding the length within the slice, we may have to mimick something like [`Vec::into_raw_parts`] return type, since I don't think its possible to slice within a const-context.

for example, this will not compile: ([playground])

```rust
struct ArrayVec<T> {
    inner: [T; 8],
    length: usize,
}

impl<T> ArrayVec<T> {
    const fn raw_parts(&self) -> &[T] {
        &self.inner[..self.length]
    }
}
```

At this point I think exposing `next_free` under a different name (length) would be less finicky, what do you think?

[playground]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ab698946fc409d92de88a5db8659700e
[`Vec::into_raw_parts`]: https://doc.rust-lang.org/std/vec/struct.Vec.html#method.into_raw_parts

Reply:
Hmm, good point. Could you tell me a bit more about your use case for `into_raw_parts`?

Reply:
FWIW I've compiled a short list of potential return types that I could think of:

* `(*const u64, usize)` a (ptr, len) tuple.
* `(&[u64], usize)` because we may want to avoid forcing the use of pointers in const.
*  an `[Option<u64>; 8]` where the unused entries are `None`
*  or up straight `[u64; 8]` which doesn't discriminate against the unused entries (but would be identifiable as `0u64`)
   * Also `&[u64]` works here which complements `from_raw_parts` just accepting a slice.

Reply:
@phil-opp sorry I didn't see your previous message, I suggested `into_raw_parts` to compliment `from_raw_parts` constructor, I don't currently have a need for it but I didn't see a reason to skip it, I can live without it fine.

Reply:
Then again, I have just realized that there's no way to actually read the GDT entries from Rust... `as_raw_parts` addresses this indirectly but I'm not sure how important it would be.

Reply:
In that case, we can just implement this function as non-const until slicing in `const` functions is implemented on the Rust side.